File: compiler/python3/host/entity_path_converter.py
"""
Helper functions for converting entity_path dictionaries to strings.
This module provides convenience wrappers around the entity_path_as_string() methods
from IDL entities.
"""

from typing import Dict, Optional, Any


def function_entity_path_to_string(func: Any, idl_definition: Any) -> str:
    """
    Convert FunctionDefinition entity_path to string.
    
    Args:
        func: FunctionDefinition instance
        idl_definition: IDLDefinition containing metaffi_guest_lib
        
    Returns:
        Entity path string in format "key1=value1,key2=value2"
    """
    return func.entity_path_as_string(idl_definition)


def method_entity_path_to_string(
    method: Any,
    idl_definition: Any,
    parent_entity_path: Optional[Dict[str, str]] = None
) -> str:
    """
    Convert MethodDefinition entity_path to string.
    
    Args:
        method: MethodDefinition instance
        idl_definition: IDLDefinition containing metaffi_guest_lib
        parent_entity_path: Optional parent class entity_path to merge
        
    Returns:
        Entity path string in format "key1=value1,key2=value2"
    """
    return method.entity_path_as_string(idl_definition, parent_entity_path)


def constructor_entity_path_to_string(ctor: Any, idl_definition: Any) -> str:
    """
    Convert ConstructorDefinition entity_path to string.
    
    Args:
        ctor: ConstructorDefinition instance
        idl_definition: IDLDefinition containing metaffi_guest_lib
        
    Returns:
        Entity path string in format "key1=value1,key2=value2"
    """
    return ctor.entity_path_as_string(idl_definition)


def global_entity_path_to_string(
    global_def: Any,
    idl_definition: Any,
    use_getter: bool = True
) -> str:
    """
    Convert GlobalDefinition entity_path to string (via getter or setter).
    
    Args:
        global_def: GlobalDefinition instance
        idl_definition: IDLDefinition containing metaffi_guest_lib
        use_getter: If True, use getter's entity_path; if False, use setter's
        
    Returns:
        Entity path string in format "key1=value1,key2=value2"
    """
    return global_def.entity_path_as_string(idl_definition, use_getter)


def field_entity_path_to_string(
    field: Any,
    idl_definition: Any,
    use_getter: bool = True
) -> str:
    """
    Convert FieldDefinition entity_path to string (via getter or setter).
    
    Args:
        field: FieldDefinition instance
        idl_definition: IDLDefinition containing metaffi_guest_lib
        use_getter: If True, use getter's entity_path; if False, use setter's
        
    Returns:
        Entity path string in format "key1=value1,key2=value2"
    """
    return field.entity_path_as_string(idl_definition, use_getter)

File: compiler/jvm/host/src/com/metaffi/compiler/host/EntityPathConverter.java
package com.metaffi.compiler.host;

import com.metaffi.idl.entities.FunctionDefinition;
import com.metaffi.idl.entities.IDLDefinition;
import com.metaffi.idl.entities.MethodDefinition;
import com.metaffi.idl.entities.ConstructorDefinition;
import com.metaffi.idl.entities.GlobalDefinition;
import com.metaffi.idl.entities.FieldDefinition;

import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

public final class EntityPathConverter {
    private EntityPathConverter() {}

    public static String functionEntityPathToString(FunctionDefinition func, IDLDefinition idlDefinition) {
        if (func == null) {
            throw new CompilerException("FunctionDefinition is required");
        }
        return entityPathToString(func.getEntityPath(), idlDefinition, null, "function", func.getName());
    }

    public static String methodEntityPathToString(MethodDefinition method, IDLDefinition idlDefinition, Map<String, String> parentEntityPath) {
        if (method == null) {
            throw new CompilerException("MethodDefinition is required");
        }
        return entityPathToString(method.getEntityPath(), idlDefinition, parentEntityPath, "method", method.getName());
    }

    public static String constructorEntityPathToString(ConstructorDefinition ctor, IDLDefinition idlDefinition) {
        if (ctor == null) {
            throw new CompilerException("ConstructorDefinition is required");
        }
        return entityPathToString(ctor.getEntityPath(), idlDefinition, null, "constructor", ctor.getName());
    }

    public static String globalEntityPathToString(GlobalDefinition globalDef, IDLDefinition idlDefinition, boolean useGetter) {
        if (globalDef == null) {
            throw new CompilerException("GlobalDefinition is required");
        }
        if (useGetter) {
            if (globalDef.getGetter() == null) {
                throw new CompilerException("Global " + globalDef.getName() + " has no getter");
            }
            return functionEntityPathToString(globalDef.getGetter(), idlDefinition);
        }
        if (globalDef.getSetter() == null) {
            throw new CompilerException("Global " + globalDef.getName() + " has no setter");
        }
        return functionEntityPathToString(globalDef.getSetter(), idlDefinition);
    }

    public static String fieldEntityPathToString(FieldDefinition field, IDLDefinition idlDefinition, Map<String, String> parentEntityPath, boolean useGetter) {
        if (field == null) {
            throw new CompilerException("FieldDefinition is required");
        }
        if (useGetter) {
            if (field.getGetter() == null) {
                throw new CompilerException("Field " + field.getName() + " has no getter");
            }
            return methodEntityPathToString(field.getGetter(), idlDefinition, parentEntityPath);
        }
        if (field.getSetter() == null) {
            throw new CompilerException("Field " + field.getName() + " has no setter");
        }
        return methodEntityPathToString(field.getSetter(), idlDefinition, parentEntityPath);
    }

    private static String entityPathToString(
        Map<String, String> entityPath,
        IDLDefinition idlDefinition,
        Map<String, String> parentEntityPath,
        String entityKind,
        String entityName) {

        if (idlDefinition == null) {
            throw new CompilerException("IDLDefinition is required");
        }
        if (entityPath == null || entityPath.isEmpty()) {
            throw new CompilerException("" + entityKind + " " + entityName + " missing entity_path");
        }

        if (entityPath.containsKey("entity_path")) {
            String raw = entityPath.get("entity_path");
            if (raw == null || raw.isEmpty()) {
                throw new CompilerException("" + entityKind + " " + entityName + " has empty entity_path");
            }
            return raw;
        }

        Set<String> keys = new TreeSet<>();
        keys.addAll(entityPath.keySet());
        if (parentEntityPath != null) {
            keys.addAll(parentEntityPath.keySet());
        }
        keys.add("metaffi_guest_lib");

        StringBuilder sb = new StringBuilder();
        boolean first = true;
        for (String key : keys) {
            String value = entityPath.get(key);
            if (value == null && parentEntityPath != null) {
                value = parentEntityPath.get(key);
            }
            if (value == null) {
                if ("metaffi_guest_lib".equals(key)) {
                    value = idlDefinition.getMetaffiGuestLib();
                } else {
                    throw new CompilerException("Missing entity_path key " + key + " for " + entityKind + " " + entityName);
                }
            }

            if (!first) {
                sb.append(',');
            }
            sb.append(key).append('=').append(value);
            first = false;
        }

        return sb.toString();
    }
}
